// ecommerce/src/core/AgileGrid/AgileGrid.story.js
import React from "react";
import { storiesOf } from "@storybook/react";
import AgileGrid from "./AgileGrid";
import {
    ImageContentBoxStory,
    ImageContentBoxStory2,
    ImageContentBoxStory3,
    ImageContentBoxStory4,
    ImageContentBoxStory5,
    ImageContentBoxStoryFixed,
    ImageContentBoxStoryFixed2,
    ImageContentBoxStoryFixed3,
    ImageContentBoxStoryFixed4
} from "../../helpers/ImageContentBox/ImageContentBox.story";
import Unwrapper from "../../helpers/Unwrapper/Unwrapper";
import { Breakpoints } from "../Layout";
import {
    ProductBoxStory,
    ProductBoxStory2,
    ProductBoxStory3,
    ProductBoxStory4
} from "../../helpers/ProductBox/ProductBox.story";
import {
    ProductBoxNikeStory,
    ProductBoxNikeStory2,
    ProductBoxNikeStory3,
    ProductBoxNikeStory4
} from "../../helpers/ProductBoxNike/ProductBoxNike.story";

const _breakpoints = [
    {
        min: 0,
        mode: "slider",
        gridStyle: {
            gridAutoColumns: "90%", // slide width
            alignItems: "start"
        }
    },
    {
        min: Breakpoints.md,
        mode: "slider",
        gridStyle: {
            gridAutoColumns: "calc(50% - 10px)", // slide width
            alignItems: "start"
        }
    },
    {
        min: Breakpoints.lg,
        mode: "grid",
        gridStyle: {
            gridTemplateColumns: `repeat(4 ,1fr)`,
            alignItems: "start"
        }
    }
];
const _breakpoints2 = [
    {
        min: 0,
        mode: "slider",
        gridStyle: {
            gridAutoColumns: "100%", // slide width
            alignItems: "end"
        }
    },
    {
        min: Breakpoints.md,
        mode: "slider",
        gridStyle: {
            gridAutoColumns: "calc(50% - 10px)", // slide width
            alignItems: "end"
        }
    },
    {
        min: Breakpoints.lg,
        mode: "grid",
        gridStyle: {
            gridTemplateColumns: `repeat(6 ,1fr)`,
            alignItems: "end"
        }
    }
];
const _breakpoints3 = [
    {
        min: 0,
        mode: "grid",
        visibleItems: 4,
        gridStyle: {
            gridTemplateColumns: `repeat(2 ,1fr)`
        }
    },
    {
        min: Breakpoints.md,
        mode: "grid",
        visibleItems: 4,
        gridStyle: {
            gridTemplateColumns: `repeat(4 ,1fr)`
        }
    },
    {
        min: Breakpoints.lg,
        mode: "grid",
        gridStyle: {
            gridTemplateColumns: `repeat(4 ,1fr)`,
            alignItems: "end"
        }
    }
];

const _breakpoints4 = [
    {
        min: 0,
        mode: "slider",
        snap: "center",
        gridStyle: {
            gridAutoColumns: "90%" // slide width
        }
    },
    {
        min: Breakpoints.md,
        mode: "slider",
        snap: "center",
        gridStyle: {
            gridAutoColumns: "90%" // slide width
        }
    },
    {
        min: Breakpoints.lg,
        mode: "slider",
        includeGridMargins: true,
        snap: "start",
        gridStyle: {
            gridAutoColumns: `calc((100% - 3 * 36px) / 4)` // slide width
        }
    }
];
const _breakpoints5 = [
    {
        min: 0,
        mode: "grid",
        gridStyle: {
            gridTemplateColumns: `repeat(2 ,1fr)`
        }
    },
    {
        min: Breakpoints.md,
        mode: "grid",
        gridStyle: {
            gridTemplateColumns: `repeat(4 ,1fr)`
        }
    },
    {
        min: Breakpoints.lg,
        mode: "grid",
        gridStyle: {
            gridTemplateColumns: `repeat(4 ,1fr)`
        }
    }
];
storiesOf("AgileGrid").add("default", () => (
    <div>
        <Unwrapper componentName={"AgileGrid"} storyName={"default"} />
        <AgileGrid breakpoints={_breakpoints}>
            {ImageContentBoxStory}
            {ImageContentBoxStory2}
            {ImageContentBoxStory3}
            {ImageContentBoxStory4}
            {ImageContentBoxStory}
            {ImageContentBoxStory4}
            {ImageContentBoxStory3}
            {ImageContentBoxStory2}
            {ImageContentBoxStory3}
            {ImageContentBoxStory4}
            {ImageContentBoxStory5}
            {ImageContentBoxStory}
            {ImageContentBoxStory4}
            {ImageContentBoxStory}
            {ImageContentBoxStory4}
            {ImageContentBoxStory3}
            {ImageContentBoxStory2}
            {ImageContentBoxStory3}
            {ImageContentBoxStory4}
            {ImageContentBoxStory}
            {ImageContentBoxStory4}
        </AgileGrid>
    </div>
));
storiesOf("AgileGrid").add("equal items content height", () => (
    <div>
        <Unwrapper
            componentName={"AgileGrid"}
            storyName={"equal%20items%20content%20height"}
        />
        <AgileGrid breakpoints={_breakpoints2}>
            {ImageContentBoxStoryFixed}
            {ImageContentBoxStoryFixed2}
            {ImageContentBoxStoryFixed3}
            {ImageContentBoxStoryFixed4}

            {ImageContentBoxStoryFixed4}
            {ImageContentBoxStoryFixed2}
            {ImageContentBoxStoryFixed}
            {ImageContentBoxStoryFixed3}

            {ImageContentBoxStoryFixed}
            {ImageContentBoxStoryFixed2}
            {ImageContentBoxStoryFixed3}
            {ImageContentBoxStoryFixed4}
            {ImageContentBoxStoryFixed3}
            {ImageContentBoxStoryFixed4}
            {ImageContentBoxStoryFixed2}
            {ImageContentBoxStoryFixed}
        </AgileGrid>
    </div>
));
storiesOf("AgileGrid").add("visibleItems", () => (
    <div>
        <Unwrapper componentName={"AgileGrid"} storyName={"visibleItems"} />
        <AgileGrid breakpoints={_breakpoints3}>
            {ImageContentBoxStoryFixed}
            {ImageContentBoxStoryFixed2}
            {ImageContentBoxStoryFixed3}
            {ImageContentBoxStoryFixed4}

            {ImageContentBoxStoryFixed4}
            {ImageContentBoxStoryFixed2}
            {ImageContentBoxStoryFixed}
            {ImageContentBoxStoryFixed3}

            {ImageContentBoxStoryFixed}
            {ImageContentBoxStoryFixed2}
            {ImageContentBoxStoryFixed3}
            {ImageContentBoxStoryFixed4}
        </AgileGrid>
    </div>
));
storiesOf("AgileGrid").add("sliders", () => (
    <div>
        <Unwrapper componentName={"AgileGrid"} storyName={"sliders"} />
        <AgileGrid breakpoints={_breakpoints4}>
            {ImageContentBoxStoryFixed}
            {ImageContentBoxStoryFixed2}
            {ImageContentBoxStoryFixed3}
            {ImageContentBoxStoryFixed4}

            {ImageContentBoxStoryFixed4}
            {ImageContentBoxStoryFixed2}
            {ImageContentBoxStoryFixed}
            {ImageContentBoxStoryFixed3}

            {ImageContentBoxStoryFixed}
            {ImageContentBoxStoryFixed2}
            {ImageContentBoxStoryFixed3}
            {ImageContentBoxStoryFixed4}
        </AgileGrid>
    </div>
));
const AgileGridStoryNike = (
    <AgileGrid breakpoints={_breakpoints}>
        {ProductBoxNikeStory}
        {ProductBoxNikeStory2}
        {ProductBoxNikeStory3}
        {ProductBoxNikeStory4}
    </AgileGrid>
);
storiesOf("AgileGrid").add("nike", () => (
    <div>
        <Unwrapper componentName={"AgileGrid"} storyName={"nike"} />
        {AgileGridStoryNike}
    </div>
));
const AgileGridStory = (
    <div>
        <div className={"AgileGrid__title"}>You may also like</div>
        <AgileGrid breakpoints={_breakpoints5}>
            {ProductBoxStory}
            {ProductBoxStory2}
            {ProductBoxStory3}
            {ProductBoxStory4}
        </AgileGrid>
    </div>
);
storiesOf("AgileGrid").add("products", () => AgileGridStory);
export { AgileGridStory, AgileGridStoryNike };
