// ecommerce/src/helpers/ProductMetaNike/ProductVariantsNike/ProductVariantsNike.story.js
import React from "react";
import { storiesOf } from "@storybook/react";
import ProductVariantsNike from "./ProductVariantsNike";

const ProductVariantsNikeStory = [
    {
        isActive: false,
        image: {
            source: "imgix-ratio",
            path:
                "https://s3.eu-central-1.amazonaws.com/ratio-assets/Storefront/nike/1-2.jpg",
            w: 1728,
            h: 2160,
            alt: ""
        }
    },
    {
        isActive: true,
        image: {
            source: "imgix-ratio",
            path:
                "https://s3.eu-central-1.amazonaws.com/ratio-assets/Storefront/nike/1-1.jpg",
            w: 1728,
            h: 2160,
            alt: ""
        }
    },
    {
        isActive: false,
        image: {
            source: "imgix-ratio",
            path:
                "https://s3.eu-central-1.amazonaws.com/ratio-assets/Storefront/nike/1-3.jpg",
            w: 1728,
            h: 2160,
            alt: ""
        }
    },
    {
        isActive: false,
        image: {
            source: "imgix-ratio",
            path:
                "https://s3.eu-central-1.amazonaws.com/ratio-assets/Storefront/nike/1-4.jpg",
            w: 1728,
            h: 2160,
            alt: ""
        }
    }
];

storiesOf("ProductVariantsNike").add("default", () => (
    <div>
        <ProductVariantsNike variants={ProductVariantsNikeStory} />
    </div>
));

export default ProductVariantsNikeStory;
