// ecommerce/src/helpers/ProductBoxNike/ProductBoxNike.js
import React from "react";
import PropTypes from "prop-types";
import "./ProductBoxNike.scss";
import LazyAssetWrapper from "../../core/LazyAssetWrapper/LazyAssetWrapper";

import InlineSVG from "svg-inline-react";
import HeartIcon from "../../svg/outline-favorite_border-18px.svg";

class ProductBoxNike extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            hoveredTriggerIndex: -1
        };
    }
    handleHover(i) {
        this.setState({
            hoveredTriggerIndex: i
        });
    }

    handleHoverLeave() {
        this.setState({
            hoveredTriggerIndex: -1
        });
    }
    render() {
        return (
            <div className="ProductBoxNike">
                <a href={this.props.link} className="ProductBoxNike__images">
                    <LazyAssetWrapper
                        className={
                            "ProductBoxNike__image ProductBoxNike__image--main"
                        }
                        mode={"natural"}
                        crop={"square"}
                        sizes={`500px`}
                        image={this.props.image}
                    />
                    <LazyAssetWrapper
                        className={
                            "ProductBoxNike__image ProductBoxNike__image--onHover"
                        }
                        mode={"natural"}
                        crop={"square"}
                        sizes={`500px`}
                        image={this.props.imageOnHover}
                    />
                    {this.props.variantsImages && (
                        <div className="ProductBoxNike__variants">
                            {this.props.variantsImages.map((image, index) => {
                                let classes = [
                                    "ProductBoxNike__image ProductBoxNike__image--alternate"
                                ];
                                if (this.state.hoveredTriggerIndex === index) {
                                    classes.push(
                                        "ProductBoxNike__image--isOnTop"
                                    );
                                }
                                return (
                                    <LazyAssetWrapper
                                        className={classes.join(" ")}
                                        mode={"natural"}
                                        crop={"square"}
                                        sizes={`500px`}
                                        key={index}
                                        image={image}
                                    />
                                );
                            })}
                        </div>
                    )}
                </a>

                <div className="ProductBoxNike__content">
                    <div className="ProductBoxNike__contentHead">
                        <div className="ProductBoxNike__titleRow">
                            <div className="ProductBoxNike__title">
                                {this.props.title}
                            </div>
                            <div className="ProductBoxNike__heart">
                                <InlineSVG src={HeartIcon} />️
                            </div>
                        </div>
                    </div>
                    <div className="ProductBoxNike__label">
                        {this.props.label}
                    </div>
                    <div className="ProductBoxNike__priceRow">
                        <div className="ProductBoxNike__price">
                            {this.props.price}
                        </div>
                        {this.props.priceBefore && (
                            <div className="ProductBoxNike__price ProductBoxNike__price--before">
                                {this.props.priceBefore}
                            </div>
                        )}
                    </div>
                    {this.props.colors && (
                        <div className="ProductBoxNike__variantsTriggers">
                            {this.props.colors.map((color, index) => {
                                return (
                                    <div
                                        className={"ProductBoxNike__trigger"}
                                        key={index}
                                        onMouseEnter={this.handleHover.bind(
                                            this,
                                            index
                                        )}
                                        onMouseLeave={this.handleHoverLeave.bind(
                                            this,
                                            index
                                        )}
                                    >
                                        <div
                                            className="ProductBoxNike__triggerInner"
                                            style={{ background: color }}
                                        />
                                    </div>
                                );
                            })}
                        </div>
                    )}
                </div>
            </div>
        );
    }
}
ProductBoxNike.propTypes = {
    image: PropTypes.object.isRequired,
    imageOnHover: PropTypes.object,
    variantsImages: PropTypes.array,
    label: PropTypes.string,
    title: PropTypes.string.isRequired,
    colors: PropTypes.array,
    link: PropTypes.string.isRequired,
    price: PropTypes.string.isRequired,
    priceBefore: PropTypes.string
};

export default ProductBoxNike;
