// ecommerce/src/helpers/ProductBoxNike/ProductBoxNike.story.js
import React from "react";
import { storiesOf } from "@storybook/react";
import ProductBoxNike from "./ProductBoxNike";

const ProductBoxNikeStory = (
    <ProductBoxNike
        title={"Nike Zoom Pegasus Turbo"}
        // colors={["#e6d4c8", "#464c58", "#d5d1c6"]}
        label={"Men's Running Shoe"}
        link={"#"}
        price={"€79"}
        image={{
            source: "imgix-ratio",
            path:
                "https://s3.eu-central-1.amazonaws.com/ratio-assets/Storefront/nike/1-1.jpg",
            w: 1728,
            h: 2160,
            alt: ""
        }}
        imageOnHover={{
            source: "imgix-ratio",
            path:
                "https://s3.eu-central-1.amazonaws.com/ratio-assets/Storefront/nike/2.jpg",
            w: 1728,
            h: 2160,
            alt: ""
        }}
    />
);
const ProductBoxNikeStory2 = (
    <ProductBoxNike
        title={"Nike Zoom Pegasus Turbo"}
        // colors={["#e6d4c8", "#464c58", "#d5d1c6"]}
        label={"Men's Running Shoe"}
        link={"#"}
        price={"€79"}
        image={{
            source: "imgix-ratio",
            path:
                "https://s3.eu-central-1.amazonaws.com/ratio-assets/Storefront/nike/1-2.jpg",
            w: 1728,
            h: 2160,
            alt: ""
        }}
        imageOnHover={{
            source: "imgix-ratio",
            path:
                "https://s3.eu-central-1.amazonaws.com/ratio-assets/Storefront/nike/2.jpg",
            w: 1728,
            h: 2160,
            alt: ""
        }}
    />
);
const ProductBoxNikeStory3 = (
    <ProductBoxNike
        title={"Nike Zoom Pegasus Turbo"}
        // colors={["#e6d4c8", "#464c58", "#d5d1c6"]}
        label={"Men's Running Shoe"}
        link={"#"}
        price={"€79"}
        image={{
            source: "imgix-ratio",
            path:
                "https://s3.eu-central-1.amazonaws.com/ratio-assets/Storefront/nike/1-3.jpg",
            w: 1728,
            h: 2160,
            alt: ""
        }}
        imageOnHover={{
            source: "imgix-ratio",
            path:
                "https://s3.eu-central-1.amazonaws.com/ratio-assets/Storefront/nike/2.jpg",
            w: 1728,
            h: 2160,
            alt: ""
        }}
    />
);
const ProductBoxNikeStory4 = (
    <ProductBoxNike
        title={"Nike Zoom Pegasus Turbo"}
        // colors={["#e6d4c8", "#464c58", "#d5d1c6"]}
        label={"Men's Running Shoe"}
        link={"#"}
        price={"€79"}
        image={{
            source: "imgix-ratio",
            path:
                "https://s3.eu-central-1.amazonaws.com/ratio-assets/Storefront/nike/1-4.jpg",
            w: 1728,
            h: 2160,
            alt: ""
        }}
        imageOnHover={{
            source: "imgix-ratio",
            path:
                "https://s3.eu-central-1.amazonaws.com/ratio-assets/Storefront/nike/2.jpg",
            w: 1728,
            h: 2160,
            alt: ""
        }}
    />
);
storiesOf("ProductBoxNike").add("default", () => ProductBoxNikeStory);

export {
    ProductBoxNikeStory,
    ProductBoxNikeStory2,
    ProductBoxNikeStory3,
    ProductBoxNikeStory4
};
