// ecommerce/src/helpers/Test/Test.js
import React from "react";
import "./Test.scss";
import InlineSVG from "svg-inline-react";

import PropTypes from "prop-types";

import PlayIcon from "../../svg/play.svg";

import { ReactSimpleSwiper } from "simpleswiper";

import LazyAssetWrapper from "../../core/LazyAssetWrapper/LazyAssetWrapper";

class Test extends React.Component {
    constructor(props) {
        super(props);

        this.slider = React.createRef();
    }

    componentDidMount() {
        this.slider.current.layout();
    }

    render() {
        return (
            <div className={"Test"}>
                <h1>{this.props.title}</h1>
                <p>
                    {this.props.text} <InlineSVG src={PlayIcon} raw />
                </p>
                <p>{this.props.secondaryText}</p>

                <LazyAssetWrapper
                    className={"AboutPeople__photo AboutPeople__photoSecondary"}
                    mode={"natural"}
                    crop={"natural"}
                    sizes={`100vw`}
                    image={{
                        source: "imgix-ratio",
                        path:
                            "https://s3.eu-central-1.amazonaws.com/ratio-assets/Jit+site/about-people/Tomasz_Klepacki_hover.jpg",
                        w: 1075,
                        h: 1613,
                        alt: "Tomasz Klepacki"
                    }}
                    placeholder={null}
                    backgroundColor={"white"}
                />

                <div>
                    dupa<p>cześć</p>
                </div>

                <ReactSimpleSwiper
                    containerStyles={{
                        width: "800px",
                        position: "relative",
                        border: "1px solid black",
                        overflow: "hidden"
                    }}
                    slideSize={() => 500}
                    rightOffset={() => 100}
                    leftOffset={() => 50}
                    slideSnapOffset={() => 50}
                    slideMargin={() => 50}
                    infinite={false}
                    ref={this.slider}
                >
                    <div
                        className="slide"
                        style={{ backgroundColor: "red", height: "500px" }}
                    >
                        1
                    </div>
                    <div
                        className="slide"
                        style={{ backgroundColor: "red", height: "500px" }}
                    >
                        2
                    </div>
                    <div
                        className="slide"
                        style={{ backgroundColor: "red", height: "500px" }}
                    >
                        3
                    </div>
                    <div
                        className="slide"
                        style={{ backgroundColor: "red", height: "500px" }}
                    >
                        4
                    </div>
                </ReactSimpleSwiper>
            </div>
        );
    }
}

Test.propTypes = {
    title: PropTypes.string.isRequired,
    text: PropTypes.string.isRequired,
    secondaryText: PropTypes.string.isRequired
};

export default Test;
