// ecommerce/src/helpers/ProductBox/ProductBox.js
import React from "react";
import PropTypes from "prop-types";
import "./ProductBox.scss";
import LazyAssetWrapper from "../../core/LazyAssetWrapper/LazyAssetWrapper";

import InlineSVG from "svg-inline-react";
import HeartIcon from "../../svg/outline-favorite_border-18px.svg";

class ProductBox extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            hoveredTriggerIndex: -1
        };
    }
    handleHover(i) {
        this.setState({
            hoveredTriggerIndex: i
        });
    }

    handleHoverLeave() {
        this.setState({
            hoveredTriggerIndex: -1
        });
    }
    render() {
        return (
            <div className="ProductBox">
                <a href={this.props.link} className="ProductBox__images">
                    <LazyAssetWrapper
                        className={"ProductBox__image ProductBox__image--main"}
                        mode={"natural"}
                        crop={"natural"}
                        sizes={`500px`}
                        image={this.props.image}
                    />
                    <LazyAssetWrapper
                        className={
                            "ProductBox__image ProductBox__image--onHover"
                        }
                        mode={"natural"}
                        crop={"natural"}
                        sizes={`500px`}
                        image={this.props.imageOnHover}
                    />
                    <div className="ProductBox__variants">
                        {this.props.variantsImages.map((image, index) => {
                            let classes = [
                                "ProductBox__image ProductBox__image--alternate"
                            ];
                            if (this.state.hoveredTriggerIndex === index) {
                                classes.push("ProductBox__image--isOnTop");
                            }
                            return (
                                <LazyAssetWrapper
                                    className={classes.join(" ")}
                                    mode={"natural"}
                                    crop={"natural"}
                                    sizes={`500px`}
                                    key={index}
                                    image={image}
                                />
                            );
                        })}
                    </div>
                </a>

                <div className="ProductBox__content">
                    <div className="ProductBox__contentHead">
                        <div className="ProductBox__titleRow">
                            <div className="ProductBox__title">
                                {this.props.title}
                            </div>
                            <div className="ProductBox__heart">
                                <InlineSVG src={HeartIcon} />️
                            </div>
                        </div>
                        <div className="ProductBox__priceRow">
                            <div className="ProductBox__price">
                                {this.props.price}
                            </div>
                            {this.props.priceBefore && (
                                <div className="ProductBox__price ProductBox__price--before">
                                    {this.props.priceBefore}
                                </div>
                            )}
                        </div>
                    </div>
                    <div className="ProductBox__label">{this.props.label}</div>
                    <div className="ProductBox__variantsTriggers">
                        {this.props.colors.map((color, index) => {
                            return (
                                <div
                                    className={"ProductBox__trigger"}
                                    key={index}
                                    onMouseEnter={this.handleHover.bind(
                                        this,
                                        index
                                    )}
                                    onMouseLeave={this.handleHoverLeave.bind(
                                        this,
                                        index
                                    )}
                                >
                                    <div
                                        className="ProductBox__triggerInner"
                                        style={{ background: color }}
                                    />
                                </div>
                            );
                        })}
                    </div>
                </div>
            </div>
        );
    }
}
ProductBox.propTypes = {
    image: PropTypes.object.isRequired,
    imageOnHover: PropTypes.object,
    variantsImages: PropTypes.array,
    label: PropTypes.string,
    title: PropTypes.string.isRequired,
    colors: PropTypes.array,
    link: PropTypes.string.isRequired,
    price: PropTypes.string.isRequired,
    priceBefore: PropTypes.string
};

export default ProductBox;
