// ecommerce/src/helpers/MenuBar/MenuBar.js
import React from "react";
import "./MenuBar.scss";
// import PropTypes from "prop-types";
import InlineSVG from "svg-inline-react";
import Logo from "../../svg/logo_cos.svg";
import Search from "../../svg/outline-search-24px.svg";

// import LazyAssetWrapper from "../LazyAssetWrapper/LazyAssetWrapper";

class MenuBar extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            isScrolledDown: false
        };
    }

    handleScroll() {
        if (this.state.isScrolledDown === false && window.scrollY > 80) {
            this.setState({
                isScrolledDown: true
            });
            console.log("true");
        } else {
            if (this.state.isScrolledDown && window.scrollY <= 80) {
                console.log("false");

                this.setState({
                    isScrolledDown: false
                });
            }
        }
    }
    componentDidMount() {
        window.addEventListener("scroll", this.handleScroll);
    }

    componentWillUnmount() {
        window.removeEventListener("scroll", this.handleScroll);
    }

    render() {
        let _classes = ["MenuBar"];

        if (this.state.isScrolledDown) {
            _classes.push("MenuBar--isScrolledDown");
        }

        return (
            <div className={_classes.join(" ")}>
                <div className="container">
                    <div className="MenuBar__inner">
                        <div className="MenuBar__innerLeft">
                            <div className="MenuBar__logo">
                                <a href="#">
                                    <InlineSVG src={Logo} raw />
                                </a>
                            </div>
                            <div className="MenuBar__link">
                                <a href="#">Women</a>
                            </div>
                            <div className="MenuBar__link">
                                <a href="#">Men</a>
                            </div>
                            <div className="MenuBar__link">
                                <a href="#">Kids and Baby</a>
                            </div>
                            <div className="MenuBar__link">
                                <a href="#">Gifts</a>
                            </div>
                            <div className="MenuBar__link">
                                <a href="#">COS × HAY</a>
                            </div>
                            <div className="MenuBar__link">
                                <a href="#">Explore</a>
                            </div>
                        </div>
                        <div className="MenuBar__innerRight">
                            <button
                                type={"button"}
                                className={"MenuBar__button"}
                            >
                                <InlineSVG src={Search} raw />
                            </button>
                            <button
                                type={"button"}
                                className={"MenuBar__button"}
                            >
                                Sign in
                            </button>
                            <button
                                type={"button"}
                                className={"MenuBar__button"}
                            >
                                Shipping to: Poland
                            </button>
                            <button
                                type={"button"}
                                className={"MenuBar__button"}
                            >
                                My Bag
                            </button>
                        </div>
                    </div>
                </div>
            </div>
        );
    }
}

// MenuBar.propTypes = {
// };

export default MenuBar;
