// ecommerce/src/helpers/ProductMeta/ProductVariants/ProductVariants.story.js
import React from "react";
import { storiesOf } from "@storybook/react";
import ProductVariants from "./ProductVariants";

const ProductVariantsStory = [
    {
        isActive: false,
        image: {
            source: "imgix-ratio",
            path:
                "https://s3.eu-central-1.amazonaws.com/ratio-assets/Storefront/product_2.jpg",
            w: 1128,
            h: 1692,
            alt: "Product 2"
        }
    },
    {
        isActive: false,
        image: {
            source: "imgix-ratio",
            path:
                "https://s3.eu-central-1.amazonaws.com/ratio-assets/Storefront/product_1.jpg",
            w: 1128,
            h: 1692,
            alt: "Product 1"
        }
    },
    {
        isActive: true,
        image: {
            source: "imgix-ratio",
            path:
                "https://s3.eu-central-1.amazonaws.com/ratio-assets/Storefront/product_2.jpg",
            w: 1128,
            h: 1692,
            alt: "Product 2"
        }
    }
];

storiesOf("ProductVariants").add("default", () => (
    <div>
        <ProductVariants variants={ProductVariantsStory} />
    </div>
));

export default ProductVariantsStory;
